task4/main.go

```
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+func getResponseTime(url string, timeout int) (duration int64, err error) {
+	time_start := time.Now()
+	client := http.Client{
+		Timeout: time.Duration(timeout) * time.Millisecond,
+	}
+	data, err := client.Get(url)
+	if (err != nil) {
+		return
+	}
+	if (data.StatusCode != 200) {
+		err = errors.New("StatusCode: " + strconv.Itoa(data.StatusCode))
+	}
+	duration = time.Since(time_start).Nanoseconds();
+	return
+}
+
+func printInfo(durations []int64) {
+	var min, max, summ int64;
+	if len(durations) > 0 {
+		min = durations[0]
+		max = durations[0]
+		for _, value := range durations {
+			if max < value {
+				max = value
+			}
+			if min > value {
+				min = value
+			}
+			summ += value;
+		}
+		avg := summ / int64(len(durations))
+		fmt.Println("Good requests : ", len(durations))
+		fmt.Println("Min time : ", time.Duration(min)/time.Nanosecond)
+		fmt.Println("Max time : ", time.Duration(max)/time.Nanosecond)
+		fmt.Println("Avg time : ", time.Duration(avg)/time.Nanosecond)
+	}
+}
+
+func performRequests(urls []string,timeout int,amt int) (durations [] int64,badRequests int){
+	mux := &sync.Mutex{}
+	var wg sync.WaitGroup
+	for i := 0; i < amt; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for _, url := range urls {
+				duration, err := getResponseTime(url, timeout)
+				if (err == nil) {
+					mux.Lock()
+					durations = append(durations, duration)
+					mux.Unlock()
+				} else {
+					badRequests++;
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	return
+}
+
+func main() {
+	url_list := flag.String("url_list", "", "url_list")
+	timeout := flag.Int("timeout", 0, "timeout")
+	amt := flag.Int("amt", 0, "amt")
+	flag.Parse()
+	content, _ := ioutil.ReadFile(*url_list)
```

all urls parse from flag (split by comma)
stop using files

also - never ignore errors